File: java/src/SqlInjection.java
package com.learning.vulnerable;

import java.sql.*;
import javax.servlet.http.*;

/**
 * VULNERABLE: SQL Injection (Java)
 * CodeQL should flag: java/sql-injection
 *
 * String concatenation in SQL queries with user input.
 */
public class SqlInjection extends HttpServlet {

    // BAD: String concatenation with user input
    protected void doGet(HttpServletRequest request, HttpServletResponse response) {
        String userId = request.getParameter("id");
        String query = "SELECT * FROM users WHERE id = '" + userId + "'";

        try {
            Connection conn = DriverManager.getConnection("jdbc:mysql://localhost/db");
            Statement stmt = conn.createStatement();
            ResultSet rs = stmt.executeQuery(query); // SQL injection here
            // process results...
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // GOOD: PreparedStatement with parameterized query
    protected void doGetSafe(HttpServletRequest request, HttpServletResponse response) {
        String userId = request.getParameter("id");

        try {
            Connection conn = DriverManager.getConnection("jdbc:mysql://localhost/db");
            PreparedStatement stmt = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
            stmt.setString(1, userId); // Safe — properly escaped
            ResultSet rs = stmt.executeQuery();
            // process results...
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: javascript/sql-injection.js
/**
 * VULNERABLE: SQL Injection
 * CodeQL should flag: js/sql-injection
 *
 * The user input flows directly into a SQL query without sanitization.
 * This is the #1 most common finding customers ask about.
 */
const express = require('express');
const mysql = require('mysql');
const app = express();

const db = mysql.createConnection({
  host: 'localhost',
  user: 'root',
  password: 'password',
  database: 'testdb'
});

// BAD: Direct string concatenation with user input
app.get('/user', (req, res) => {
  const userId = req.query.id;
  const query = "SELECT * FROM users WHERE id = '" + userId + "'";
  db.query(query, (err, results) => {
    if (err) return res.status(500).send(err);
    res.json(results);
  });
});

// GOOD: Parameterized query (safe)
app.get('/user-safe', (req, res) => {
  const userId = req.query.id;
  const query = "SELECT * FROM users WHERE id = ?";
  db.query(query, [userId], (err, results) => {
    if (err) return res.status(500).send(err);
    res.json(results);
  });
});

app.listen(3000);
